src/inmanta/deploy/scheduler.py

```diff
@@ -640,7 +665,7 @@ async def get_resource_intent_for_deploy(self, resource: ResourceIdStr) -> Optio
             self._deploying_latest.add(resource)
             return ResourceIntent(model_version=self._state.version, details=resource_details, dependencies=dependencies)
 
-    async def report_resource_state(
+    async def deploy_done(
```

Comment:
This method grew *a lot* in complexity. I think all pieces you added are indeed required, but I'd like to offer some suggestions to change the structure in order to get the complexity back to a readable level. Primarily, I think it gets complex now because some concerns are spread over the method. Concrete suggestions:

1. Make abstraction of the event propagation. Let's make it a private helper method that expects to be called under the lock. It really only needs to know the resource id, and whether this deploy recovered from a fail state (`send_recovery_events: bool`?). The method can then do the calculation of the concerned resources, and we lose a whole lot of lines in this method.
2. The current implementation is a mix of styles. This is a smaller issue than the other, and one I'm OK with to let slide (plus I think it will already improve by addressing 1), but it does feel a bit off to me that most of the method is quite functional, and then suddenly a mutable variable shows up that is updated in various unexpected places (`concerned_resources`). I think it can easily be inlined in the set comprehension for the events. Or if you don't like that, construct the two sets separately and simply `|` them.

Reply:
1. refactored this into the `_propagate_events` method
2. Couldn't inline it (I think) because we want to add `concerned_resources` to the dirty set but not event listeners but still mark both sets for re-deployment.